File: pointers/HasCycle.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public bool HasCycle(ListNode head) {
        // linked list
        // detect cycles
        // slow and fast pointers = tortoise and hare algorithm
        ListNode slowP = head;
        ListNode fastP = head;
        // make sure both pointers are not null
        while(fastP != null && fastP.next !=null){
            slowP = slowP.next;
            fastP = fastP.next.next;
            if(slowP == fastP) return true;
        }
        return false;      
    }
}

File: pointers/HasCycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fastP = head
        slowP = head
        while (fastP is not None and fastP.next is not None):
            slowP = slowP.next
            fastP = fastP.next.next
        if slowP == fastP:
            return True
        return False